openwisp-config/files/openwisp.agent

```diff
@@ -562,12 +563,43 @@ update_info() {
 	rm $UPDATE_INFO
 }
 
+print_error_logs() {
+	local last_update_line_no logs log_bytes max_log_bytes
+	max_log_bytes=1024
+	# Save openwisp_config related logs to a temporary file
+	logread -e openwisp >"$CONFIG_ERROR_LOG"
+	# Get the line number of the last configuration update
+	last_update_line_no=$(
+		grep "Local configuration outdated" -n \
+			"$CONFIG_ERROR_LOG" \
+			| tail -1 \
+			| cut -d : -f 1
+	)
+	# Print all logs since the last configuration update to stdout
+	# after replacing new lines with "\n"
+	logs=$(tail "$CONFIG_ERROR_LOG" -n +"$last_update_line_no" | sed 's/\n/\<br\>/g')
+	# Truncate the logs if it execeeds "max_log_bytes"
+	log_bytes=${#logs}
+	if [ "$log_bytes" -gt "$max_log_bytes" ]; then
+		logs=$(echo "$logs" | head -c "$max_log_bytes")
+		logs="$logs\n[truncated]"
+	fi
+	# Print the logs on the stdout
+	echo "$logs"
+	rm $CONFIG_ERROR_LOG
+}
+
 # report configuration status: "applied" or "error"
 report_status() {
-	local exit_code status
+	local exit_code status payload error_log
+	payload="key=$KEY&status=$1"
+	if [ "$1" = "error" ]; then
+		error_log=$(print_error_logs)
+		payload="$payload&error_reason=$error_log"
+	fi
 	# retry several times
 	for i in $(seq 1 30); do
-		$FETCH_COMMAND -i --data "key=$KEY&status=$1" "$REPORT_URL" >$STATUS_REPORT
+		$FETCH_COMMAND -i --data "$payload" "$REPORT_URL" >$STATUS_REPORT
```

Comment:
whatch out! We are sending a blob of text as a querystring parameter, we have to make sure this data is urlencoded or we can incur into issues, it seems curl provides [an easy way to do that](https://stackoverflow.com/questions/296536/how-to-urlencode-data-for-curl-command).